### PR TITLE
Split shards guide into separate GitLab and GitHub instructions

### PR DESCRIPTION
**Pull request by KCErb**
I wrote up GitLab instructions and subdivided the shards writing tutorial as described in #286.

### CODE REVIEW

**Comment by straight-shoota** [comment]
I don't think this is a good idea. Both texts are in large parts redundant. The instructions for hosting a git repository on GitHub and GitLab are not very different, so is any other Git hosting service. 

Apart from minor details, both guides differ in using Travis-CI vs Gitlab Pipelines. But that's an entirely different topic. And by no means strict relations. You can use Travis with Gitlab perfectly fine. Or host an Github and run CI in Gitlab Pipelines.

There should be only *one* guide about the concept - hosting a shard with a git hosting service. It should include specific details and examples for Github and Gitlab (potentially others could be added as well).
This would include the main sections of both guides and the *Releases* sections.

Additional guides should describe continuous integration as well as building/publishing API documentation. There are other resources as well, for example the great [blog post about Circle CI integration](https://crystal-lang.org/2018/09/04/using-circleci-2.0-for-your-crystal-projects.html) by @bcardiff.

**Comment by KCErb** [comment]
> I don't think this is a good idea

I would have appreciated a heads up over in #286 to avoid time spent on an unwelcome addition. 😤 

> Both texts are in large parts redundant.

According to this tool: https://copyleaks.com/compare-embed/compare-two-files/9aedb47a-f8f5-4816-8bc1-9d38d89aadae/6318845/1/1?key=L4sdJCv6xLJFt5hKITCc the two texts are only 12% identical.

> Apart from minor details

I disagree that the other 88% of the text I wrote consists of "minor" details.

When I was first trying to get my shard up on GitLab, I ran into a lot of difficulties reproducing the elements from the GitHub guide. This write-up condenses what was hours of reading GitLab docs and experimentation on my part into a simple guide.

> both guides differ in using Travis-CI vs Gitlab Pipelines. But that's an entirely different topic.

I'm sure I agree with you here but it doesn't change the fact that the current guide is strictly tied to the GitHub way of thinking and Travis-CI. GitLab is a vertically integrated DevOps provider, so if your guide will explain devops at all it's only symmetric to describe the github way (third party tools like travis and circle) as well as the gitlab way (their own services).

> You can use Travis with Gitlab perfectly fine.

If you want to have a public repository and pay nothing for Travis that's true. If you want a private repository and to still have access to devops (for free) ... look I'm just thinking that there's a reason GitLab exists as a separate entity from GitHub, different business model, different economics, different tool.

> There should be only one guide about the concept - hosting a shard with a git hosting service.

Do we really need a guide for how to use a host? Both guides only dedicate a handful of lines to the hosting aspect (the 12% that is identical). The rest of the guide is for all of the elements which make for a nice project but might not occur to a newcomer. Badges, releases, docs hosted on a dedicated static webserver ... these topics are the meat of this guide and what give it value.

Furthermore, splitting these out into separate guides is easier to visually parse and makes it easier to add guides for other services in the future. Simply adding a section for "specific details and examples for xxx" doesn't scale as nicely.

> Additional guides should describe continuous integration as well as building/publishing API documentation.

I understand your point here, it's nice to separate concerns in any problem-solving space, but I think the guides serve a different purpose than the rest of the docs. I think the purpose of these is to provide a tutorial-like feeling that walks you through an example project and puts all of those separate concerns together. To that end, separating these out into separate guides on hosting and CI would be a disservice to the user.

File: guides/hosting/github.md [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+
+- Create a project with the same `name` and `description` as specified in your `shard.yml`.
+
+- Add and commit everything:
+```bash
+$ git add -A && git commit -am "shard complete"
+```
+- Add the remote: (Be sure to replace `<YOUR-GITHUB-USERNAME>` and `<YOUR-REPOSITORY-NAME>` accordingly)
+
+NOTE: If you like, feel free to replace `public` with `origin`, or a remote name of your choosing.
```
Comment (RX14) [inline guides/hosting/github.md]: `origin` is the official standard, let's use it.

File: guides/hosting/github.md [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+According to [the Crystal Shards README](https://github.com/crystal-lang/shards/blob/master/README.md),
+> When libraries are installed from Git repositories, the repository is expected to have version tags following a semver-like format, prefixed with a `v`. Examples: v1.2.3, v2.0.0-rc1 or v2017.04.1
+
+Accordingly, in the input that says `tag version`, type `v0.1.0`. Make sure this matches the `version` in `shard.yml`. Title it `v0.1.0` and write a short description for the release.
+
+Click "Publish release" and you're done!
+
+You'll now notice that the GitHub Release badge has updated in your README.
+
+Follow [Semantic Versioning](http://semver.org/) and create a new release every time your push new code to `master`.
```
Comment (RX14) [inline guides/hosting/github.md]: Please clarify this: releasing every time you push isn't good advice

File: guides/hosting/github.md [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+If you haven't already, [sign up for Travis CI](https://travis-ci.org/).
+
+Insert the following markdown build badge below the description in your README.md:
+(be sure to replace `<YOUR-GITHUB-USERNAME>` and `<YOUR-REPOSITORY-NAME>` accordingly)
+```Markdown
+[![Build Status](https://travis-ci.org/<YOUR-GITHUB-USERNAME>/<YOUR-REPOSITORY-NAME>.svg?branch=master)](https://travis-ci.org/<YOUR-GITHUB-USERNAME>/<YOUR-REPOSITORY-NAME>)
+```
+Build badges are a simple way to tell people whether your Travis CI build passes.
+
+Add the following lines to your `.travis.yml`:
```
Comment (RX14) [inline guides/hosting/github.md]: Shouldn't we add these lines to the default generated `.travis.yml`?

File: guides/hosting/github.md [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+Accordingly with the outcome of this command, Travis CI will return a [build status](https://docs.travis-ci.com/user/customizing-the-build/#Breaking-the-Build) of "passed", "errored", "failed" or "canceled".
+
+
+If you want to verify that all your code has been formatted with `crystal tool format`, add a script for `crystal tool format --check`. If the code is not formatted correctly, this will [break the build](https://docs.travis-ci.com/user/for-beginners/#Breaking-the-Build) just as failing tests would.
+
+e.g.
+```YAML
+script:
+  - crystal spec
+  - crystal tool format --check
```
Comment (RX14) [inline guides/hosting/github.md]: ditto

File: guides/hosting/gitlab.md [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+
+    ```bash
+    $ git add -A && git commit -am "shard complete"
+    ```
+
+- Create a GitLab project with the same `name` and `description` as specified in your `shard.yml`.
+
+- Add the remote: (Be sure to replace `<YOUR-GITLAB-USERNAME>` and `<YOUR-REPOSITORY-NAME>` accordingly)
+
+    NOTE: If you like, feel free to replace `public` with `origin`, or a remote name of your choosing.
```
Comment (RX14) [inline guides/hosting/gitlab.md]: same as above

File: guides/hosting/gitlab.md [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+
+The result of naming this job `pages` and putting our docs in the `public` directory and specifying it as an `artifact` is that GitLab will deploy the site in that directory to the default URL `https://<YOUR-GITLAB-USERNAME>.gitlab.io/<YOUR-REPOSITORY-NAME>`.
+
+The `before_script` and `cache` keys in the file are for running the same script in every job (`shards install`) and for hanging onto the files that were created (`cache`). They're not necessary if your shard doesn't have any dependencies.
+
+If you commit the above file to your project and push, you'll trigger your first run of the new pipeline.
+
+```bash
+$ git add -A && git commit -am 'Add .gitlab-ci.yml' && git push public master
```
Comment (RX14) [inline guides/hosting/gitlab.md]: `origin`

File: guides/hosting/gitlab.md [hunk 7]
```diff
@@ -0,0 +1,157 @@
...
+or you can create the tag from the command line like so:
+
+```bash
+    $ git tag -a v0.1.0 -m "Release v0.1.0"
+```
+
+push it up
+
+```bash
+    $ git push public master --follow-tags
```
Comment (RX14) [inline guides/hosting/gitlab.md]: ditto

File: guides/hosting/gitlab.md [hunk 8]
```diff
@@ -0,0 +1,157 @@
...
+### Release Badge
+
+If you'd like you can also add a `shields.io` badge for the release. GitLab doesn't have full support for this kind of thing, and until someone adds a [version badge for gitlab](https://github.com/badges/shields/blob/master/doc/TUTORIAL.md) to shields.io, we'll have to just code in the version number in the URLs directly.
+
+- Link: `https://img.shields.io/badge/release-<VERSION>-brightgreen.svg`
+- Image: `https://img.shields.io/badge/release-<VERSION>-brightgreen.svg`
+
+where `<VERSION>` is the version number prefixed with a `v` like this: `v0.1.0`.
+
+### Mirror to GitLab
```
Comment (RX14) [inline guides/hosting/gitlab.md]: typo: `GitHub`

File: guides/hosting/gitlab.md [hunk 9]
```diff
@@ -0,0 +1,157 @@
...
+
+At the moment, crystalshards.xyz only uses the GitHub API, so if you want your library to be indexed on that service you can either:
+
+* Start a discussion about it over on their [github repo](https://github.com/zamith/crystalshards/issues) and eventually submit a Pull Request.
```
Comment (RX14) [inline guides/hosting/gitlab.md]: This is the wrong repo for crystalshards.xyz, and encouraging people to open an issue in an official tutorial is unacceptable, you should start an issue yourself and link to it if anything.

File: guides/writing_shards.md [hunk 10]
```diff
@@ -14,9 +14,9 @@ In this tutorial, we'll be making a Crystal library called _palindrome-example_.
 ### Requirements
 
 In order to release a Crystal Shard, and follow along with this tutorial, you will need the following:
-* A working installation of the [Crystal compiler](../using_the_compiler/README.md)
-* A working installation of [Git](https://git-scm.com)
-* A [GitHub](https://github.com) account
+* A working installation of the [Crystal compiler](../using_the_compiler/README.md).
+* A working installation of [Git](https://git-scm.com).
+* A [GitHub](https://github.com) or [GitLab](https://gitlab.com/) account.
```
Comment (RX14) [inline guides/writing_shards.md]: Adding full stops to the list items here should be another PR unifying typographical style of lists.

File: guides/writing_shards.md [hunk 11]
```diff
@@ -26,17 +26,16 @@ In your terminal: `crystal init lib <YOUR-SHARD-NAME>`
...
-      create  palindrome-example/shard.yml
-      create  palindrome-example/src/palindrome-example.cr
-      create  palindrome-example/src/palindrome-example/version.cr
-      create  palindrome-example/spec/spec_helper.cr
-      create  palindrome-example/spec/palindrome-example_spec.cr
+$ crystal init lib palindrome-example
+    create  palindrome-example/.gitignore
+    create  palindrome-example/.editorconfig
+    create  palindrome-example/LICENSE
+    create  palindrome-example/README.md
```
Comment (RX14) [inline guides/writing_shards.md]: Please don't change things unneccesarily

File: guides/writing_shards.md [hunk 12]
```diff
@@ -50,10 +49,9 @@ cd palindrome-example
 Then `add` & `commit` to start tracking the files with Git:
 
 ```bash
- $  git add -A
- $  git commit -am "First Commit"
-[master (root-commit) 77bad84] First Commit
- 10 files changed, 102 insertions(+)
+$ git add -A && git commit -m "First Commit"
+[master (root-commit) b86b4f7] First Commit
+ 9 files changed, 103 insertions(+)
```
Comment (RX14) [inline guides/writing_shards.md]: ditto

**Comment by straight-shoota** [comment]
@KCErb I appreciate your work, it's a great addition. It can certainly be put to use, I just suggest to change the structure. The content is great. But I think the presentation can be improved.

Also note that I don't even decide what changes are welcome, anyway.

> the two texts are only 12% identical.

The redundant parts are about creating and pushing a repository and making a release.

The majority of the non-redundant content is not strictly on the topic "Hosting a Shard" but focuses on related topics such as CI and API docs which are orthogonal to the main topic.

> I'm just thinking that there's a reason GitLab exists as a separate entity from GitHub, different business model, different economics, different tool.

Sure. But the relevant part here is how to host a shard. And that's almost identical with both services (or any other git hosting service for that matter).

> Furthermore, splitting these out into separate guides is easier to visually parse and makes it easier to add guides for other services in the future.

So, how would you, for example add a guide to Circle-CI integration?

> To that end, separating these out into separate guides on hosting and CI would be a disservice to the user.

I disagree. It's misleading to suggest that when hosting on Github, you use Travis CI and when hosting on Gitlab, you use Github Pipelines. This is strictly not the case. Git hosting and CI need to be treated separately.
The service for the user would be to explain how to host a shard on a git hosting service and then recommend specialized guides about related topics such as CI and API docs.

File: guides/hosting/github.md (hunk 3 above)
Comment (KCErb) [reply guides/hosting/github.md]: I'm not sure what to do with this comment. 😕

File: guides/hosting/github.md (hunk 2 above)
Comment (KCErb) [reply guides/hosting/github.md]: I can't clarify this, it's not my idea. I assumed since it was in the official docs that someone thinks it's a good idea and that I could learn why (if I wanted to) with a little googling ... happy to delete it.

File: guides/hosting/github.md (hunk 3 above)
Comment (bew) [reply guides/hosting/github.md]: Add the lines to https://github.com/crystal-lang/crystal/blob/master/src/compiler/crystal/tools/init/template/travis.yml.ecr which is a template auto-generated when you create a new crystal project using `crystal init ...`

**Comment by KCErb** [comment]
@straight-shoota Thanks for both your positive feedback and your constructive criticism 😀. You bring up some good points and I see now the value of separating the guides. I originally took your comment as concrete feedback on my PR, but it seems based on your most recent comment that you were just sharing your opinion. If there is something concrete I can do about your points let me know.

@RX14 Thank you also for your feedback. I implemented *most* of your recommended changes but I struggled to implement all of them because I'm not clear on what you see as the scope of the PR. On the one hand, you said

> Adding full stops to the list items here should be another PR unifying typographical style of lists.

and

> Please don't change things unneccesarily

but on the other hand you asked me to edit the GitHub guide (change `public` to `origin`, edit the suggestion to release on every push to `master`) which I consider strictly outside the scope of this PR.

I don't really want to create unneccessary friction, I just want to help! If me pushing back a little bit isn't helpful, I can just implement the other changes as you've described them verbatim and we can move on. I'd rather be a team player than "right" 😄

File: guides/hosting/github.md (hunk 3 above)
Comment (KCErb) [reply guides/hosting/github.md]: Fair enough, I guess I meant I don't know what to do in the scope of this pull request.

**Comment by RX14** [comment]
@KCErb my apologies! I didn't see that those were copied directly from the existing documentation. You are right, that should be another PR.

File: guides/hosting/github.md [hunk 13]
```diff
@@ -0,0 +1,123 @@
...
+```
+
+This tells Travis CI to generate your documentation.
+
+Next, add the following lines to your `.travis.yml`.
+(Be sure to replace all instances of `<YOUR-GITHUB-REPOSITORY-NAME>` accordingly)
+```YAML
+deploy:
+  provider: pages
+  skip_cleanup: true
```
[the remark this answers is not in the document]
Comment (KCErb) [reply guides/hosting/github.md]: If I understand the discussion here correctly, the consensus is to leave changes like this out of this PR.

[the remark this answers is not in the document]
Comment (vladfaust) [reply guides/hosting/github.md]: Ah, ok

**Comment by kalinon** [comment]
@KCErb is this still needed?

**Comment by KCErb** [comment]
In my opinion, yes, this is valuable. Those with merge authorization are certainly welcome to merge it in if they agree or let me know if the last commit was insufficient to address concerns :)

**Comment by kalinon** [comment]
@KCErb there are currently 1 change request still pending and a conflict. I also think this valuable. If we can get those resolved we can push for a new review.

**Comment by KCErb** [comment]
The change request I see is from @RX14's review. My understanding of our conversation was that they agreed that the changes were outside the scope of this PR. Specifically, because the original that I was editing 
* used "public" instead of origin
* recommended releasing on every push
* had non-ideal Travis CI instructions
it was not appropriate to fix those in this PR but they instead belong to a different one.

Perhaps @RX14 can resolve that conversation / review?

**Comment by KCErb** [comment]
I think I've resolved the conflict, it seems that the Github guide had some minor updates to Travis CI links so I moved them to the separate Github hosting page.

File: guides/hosting/gitlab.md [hunk 14]
```diff
@@ -0,0 +1,145 @@
...
+
+- Link: `https://<YOUR-GITLAB-USERNAME>.gitlab.io/<YOUR-REPOSITORY-NAME>`
+- Image: `https://img.shields.io/badge/docs-available-brightgreen.svg`
+
+
+## Releases
+
+A release is just a special commit in your history with a name (see [tagging](https://git-scm.com/book/en/v2/Git-Basics-Tagging)).
+
+According to [the Crystal Shards README](https://gitlab.com/crystal-lang/shards/blob/master/README.md),
```
Comment (kalinon) [inline guides/hosting/gitlab.md]: Think this needs to be: https://github.com/crystal-lang/shards/blob/master/README.md

Comment (kalinon) [inline guides/hosting/gitlab.md]:
```suggestion
According to [the Crystal Shards README](https://github.com/crystal-lang/shards/blob/master/README.md),
```

Comment (kalinon) [reply guides/hosting/gitlab.md]: This should fix failing build.

**Comment by KCErb** [comment]
Ah, nice catch. Thanks!
